aten/src/ATen/native/cuda/ForeachFunctors.cuh

```diff
@@ -18,7 +18,7 @@ inline void increment_version(TensorList tensors) {
 }
 
 // Initializes args and checks if all args are aligned
-template <int depth, typename T>
+template <int depth, typename T, template <int> class TensorListMetadata>
```

Comment:
Since we are presumably increasing the amount of specialization, is there any estimate of how much this will increase the binary size per compute-capability?

Reply:
Good call. I think this would double the size taken by foreach functions, which sounds bad.

It's possible to avoid the specialization, but I'm not sure if it's possible to do so without compromising the fast-path (e.g. if we use the same integer types in both the static and dynamic arg struct, we'd be decreasing the capacity of the static arg struct).

I'll measure the size impact and think more about this.

Reply:
Not saying that this should be a blocker for the PR, but it would be useful to know (especially for future changes or updates to this code) what impact increased specialization would have

Reply:
I measured the binary size. It was 219MB -> 224MB (a 2.28% increase).

I reworked the small buffer optimization mechansim in #119764 so the optimization in this PR can be applied without increasing binary size. I'd appreciate it if you can take a look when you get a chance.